### PR TITLE
Refactoring the PR parameter

### PR DESCRIPTION
**Pull request by srbarrios**
## What does this PR do?

* My last PR doesn't catch the case of re-run a tests, as we were running re-trigger by PR, exactly like unreviewed_pr method but without the iteration through all the PRs. For some reason, the unittests didn't catch it, I would need to improve them at some point...

What we need is to check exactly the same steps, but only for the specified PR, so:
- b.pr_equal_specific_branch?(pr)
- b.retrigger_check(pr)
- b.unreviewed_new_pr?(pr, comm_st)
- b.reviewed_pr?(comm_st, pr)

* Also, we change the main loop, so now we don't break the loop with first test passed on first PR pending, instead we iterate through all the known pending PRs and run the tests if possible.

* In the other hand, we found a corner case with contexts that you want to re-trigger but you don't want to have a comment or checkbox on the PR. For that case, we created the new parameter `--force_test` that will directly run the test for that PR number without check other filters.

### CODE REVIEW

File: gitarro.rb [hunk 1]
```diff
@@ -21,14 +25,18 @@
...
 
+  # force test without any filter or check
+  next if b.force_run_test(pr)
+
   # retrigger if magic word found
   b.retrigger_check(pr)
   # 0) do test for unreviewed pr
-  break if b.unreviewed_new_pr?(pr, comm_st)
+  next if b.unreviewed_new_pr?(pr, comm_st)
+
```
Comment (MalloZup) [inline gitarro.rb]: Why are you using next instead of break?

File: doc/BASICS.md [hunk 2]
```diff
@@ -30,6 +30,7 @@ Optional options:
     -f, --file '.py'                 pr_file type to filter/trigger the test against: .py, .rb
     -d, --description 'DESCRIPTION'  Test decription
     -C, --check                      Check if there is any PR requiring a test, but do not run it.
+    -F, --force_test                 Force to run a test, even if is not marked to be re-triggered.
```
Comment (MalloZup) [inline doc/BASICS.md]: I'm thinking in sake of minimalism if we could embed this option into the PR option, so we could avoid to add a new one. Imho having to much options can create a matrix of different variables and their combination that lead to bugs. If we have 8 flags we have more possibilities then 4. Number are take randomly here

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,10 @@
 # Changelog Gitarro
 
+## 0.1.86
+
+- triggered_by_pr_number method removed, we will use --PR parameter only to don't iterate through all open PRs
```
Comment (MalloZup) [inline CHANGELOG.md]: For the removal imho we can use a line. Split this in 2 line since are 2 Concepts. Removal and addition

File: lib/gitarro/backend.rb [hunk 4]
```diff
@@ -192,6 +192,15 @@ def open_newer_prs
     prs
   end
 
+  # public forcing to run the test
```
Comment (MalloZup) [inline lib/gitarro/backend.rb]: Embeding the force test to pr option is imho better because now if you run the force tests, without the pr number option, it, will always testing the 1st pr which doesn't make any value

File: gitarro.rb (hunk 1 above)
Comment (srbarrios) [reply gitarro.rb]: Oh yes, I didn't explained that.
I don't saw the reason why we cut the loop and not iterate through all the pending PRs of the list, running the tests in the same execution.
It make more sense than run the script n times, one for each PR with pending tests.
What do you think?

File: doc/BASICS.md (hunk 2 above)
Comment (srbarrios) [reply doc/BASICS.md]: No we can't.
PR specify a concrete PR to review, but will not run the test if we don't have marked it for re-trigger.
But we can have special cases, like the one we have in my project, where a merged_enabled context must run even if we don't have a checkbox or a comment for it, as is kind of a hidden context.

File: lib/gitarro/backend.rb (hunk 4 above)
Comment (srbarrios) [reply lib/gitarro/backend.rb]: Good catch. I added the missing condition.
`return unless @force_test and @pr_number.any?`

File: doc/BASICS.md (hunk 2 above)
Comment (MalloZup) [reply doc/BASICS.md]: ok thx

File: gitarro.rb (hunk 1 above)
Comment (MalloZup) [reply gitarro.rb]: We are doing this because:

1 Jenkins JOB = 1 test run.

If we run all the tests at same time, we will break this rule and we don't have 1 Build URL pro JOB/tests. SO we can't append the URL to the result of a single PR in GitHub.
We will run all the tests in a single Jenkins job with all the LOGs togheter, and run all the tests in same execution has other disadvantages. You have 30 min execution and imagine the worker dies for some reason. It is better to have small operation which are independent from each others.

It is better here to don't break the logic because we will add to much unknown for a PR which is sensed to remove and add

Comment (srbarrios) [reply gitarro.rb]: OK, I'll revert it.
Is true that mixing PR reports is bad.

**Comment by MalloZup** [comment]
thx @srbarrios for this. I will re-review it later during this week. We might also try it live in prod before merging it, just for the case.

**Comment by srbarrios** [comment]
> thx @srbarrios for this  . I will re-review it later during this week. We might also try it live in prod before merging it (when is pretty done), just for the case if we aren't overseeing things.

Agree. Let's have it few days to be sure all is fine, before create the gem.
As commented, in my project we are checking out the code of that PR, so we are already using it.

**Comment by juliogonzalez** [comment]
It seems to me something is broken: https://github.com/uyuni-project/uyuni/pull/950 is waiting for tests, but https://ci.suse.de/view/Manager/view/Uyuni-PRs/job/uyuni-prs-check/12267/console didn't detect it.

Am I missing something?

**Comment by srbarrios** [comment]
> It seems to me something is broken: [uyuni-project/uyuni#950](https://github.com/uyuni-project/uyuni/pull/950) is waiting for tests, but https://ci.suse.de/view/Manager/view/Uyuni-PRs/job/uyuni-prs-check/12267/console didn't detect it.
> 
> Am I missing something?

Indeed, fixed, was related with those breaks in the main loop.

**Comment by juliogonzalez** [comment]
It seems we still have problem with the PR.

This happens when running merge_enabled.

+ ruby gitarro.rb -r uyuni-project/uyuni --force_test -c merge_enabled -d 'Merging enabled (requires changelog_test passing)' -u https://ci.suse.de/job/uyuni-prs-changelog_test-pipeline/50339/ -g /home/jenkins/jenkins-build/workspace/uyuni-prs-changelog_test-pipeline -P 950 -t /usr/bin/true
/home/jenkins/jenkins-build/workspace/uyuni-prs-changelog_test-pipeline/lib/gitarro/backend.rb:197:in `force_run_test': undefined method `any?' for 950:Fixnum (NoMethodError)
	from gitarro.rb:29:in `block in <main>'
	from gitarro.rb:20:in `each'
	from gitarro.rb:20:in `<main>'

So PRs can't be merged.

**Comment by srbarrios** [comment]
> It seems we still have problem with the PR.
> 
> This happens when running merge_enabled.
> 
> * ruby gitarro.rb -r uyuni-project/uyuni --force_test -c merge_enabled -d 'Merging enabled (requires changelog_test passing)' -u https://ci.suse.de/job/uyuni-prs-changelog_test-pipeline/50339/ -g /home/jenkins/jenkins-build/workspace/uyuni-prs-changelog_test-pipeline -P 950 -t /usr/bin/true
>   /home/jenkins/jenkins-build/workspace/uyuni-prs-changelog_test-pipeline/lib/gitarro/backend.rb:197:in `force_run_test': undefined method `any?' for 950:Fixnum (NoMethodError)
>   from gitarro.rb:29:in `block in <main>' from gitarro.rb:20:in `each'
>   from gitarro.rb:20:in `'
> 
> So PRs can't be merged.

Fixed

File: gitarro.rb [hunk 5]
```diff
@@ -8,9 +8,13 @@
 require_relative 'lib/gitarro/backend'
 
 b = Backend.new
-exit 0 if b.triggered_by_pr_number?
 
-prs = b.open_newer_prs
+prs = if b.options[:pr_number].nil?
```
Comment (MalloZup) [inline gitarro.rb]: can we create an helper function for this?
It can resided in same file.

Or we can use guard clause for avoding if/else cognitive load.

**Comment by srbarrios** [comment]
> we need either some unit-test or rake tests. this would be really good to have both.

I only added one rake test, I consider not enough and I'd like to improve the overall coverage.
But if we can merge that PR as it is, it will help to our team.

Could we consider it @MalloZup ?

We are using that PR in our CI for about 3 weeks, and it looks stable.

**Comment by MalloZup** [comment]
@srbarrios Helping your team is great. :) 
However, we shouldn't justify the lack of tests when adding a feature with good motivation :grin:  

I think we all want to help but on the other side we have our duties as developers for whatever codebase.

Not adding a tests for a new feature, it has costs and it is not a good practice when people refactor things or when people want to add a new feature/removal code.

The fact that you are using it on CI for 3 weeks it doesn't maybe not cover all the possible scenarios.

But OK, for this PR I take your word "I'd like to improve the overall coverage" for this time I'm making an exception and trust the team.

Note that I always pretend people to do test for a XX feature, regardless codebase/projects.

The same logic I apply to myself when codebases start to become more the 100 line of code or used it.  So isn't that I'm just preaching around, but I have also my skin in the game.. :grin: 

Feel free to add a follow-up card etc, for any help regarding the tests  feel free to ask.

THX for this. :+1:

**Comment by srbarrios** [comment]
> @srbarrios Helping your team is great. :)
> However, we shouldn't justify the lack of tests when adding a feature with good motivation 
> 
> I think we all want to help but on the other side we have our duties as developers for whatever codebase.
> 
> Not adding a tests for a new feature, it has costs and it is not a good practice when people refactor things or when people want to add a new feature/removal code.
> 
> The fact that you are using it on CI for 3 weeks it doesn't maybe not cover all the possible scenarios.
> 
> But OK, for this PR I take your word "I'd like to improve the overall coverage" for this time I'm making an exception and trust the team.
> 
> Note that I always pretend people to do test for a XX feature, regardless codebase/projects.
> 
> The same logic I apply to myself when codebases start to become more the 100 line of code or used by a share team. So isn't that I'm just preaching around, but I have also my skin in the game.. 
> 
> Feel free to add a follow-up card etc, and if you need any help regarding the tests , etc feel free to ask as always 
> 
> THX for this.

Sure. I created a card to address it. To add more tests related with the new parameters, we probably need to refactor some things first https://github.com/openSUSE/gitarro/issues/172

**Comment by MalloZup** [comment]
@juliogonzalez feel free to create the gems! I don't mind if Oscar is added or want to be added in the release process, you can add it so he can create the gems if he wish! thx

**Comment by srbarrios** [comment]
Thanks @MalloZup, we'll work on that next week :)
